### PR TITLE
Added the Set-FileAttributes

### PR DESCRIPTION
**Pull request by chrisjohnson1**
##### SUMMARY

Added functionally inside of the win_file.ps1 module in order to make modifications to file attributes.  The following attributes, at this time, can be modified: Read Only, Hidden, System, and Archive.

Additionally, users of the function can set the file to Normal which resets all attributes of a windows file.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

lib/ansible/modules/windows/win_file.ps1

##### ANSIBLE VERSION
```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.6 (default, Jun 22 2015, 17:58:13) [GCC 4.8.2]
```

##### ADDITIONAL INFORMATION
```
The Set-FileAttribute function has an attribute parser built into itself for understanding the intention of the user.  Users can use +h or h to turn on the hidden attribute while using -h to remove the bit which makes the file hidden.  The same applies for: a,h,s,r,n - archive, hidden, system, readonly, normal respectively. 

Moreover, touch state was updated to be able to handle read only files.  Additionally, other CMD-let's used in the module assumed files weren't hidden.
```

### CODE REVIEW

File: lib/ansible/modules/windows/win_file.ps1 [hunk 1]
```diff
@@ -25,6 +25,7 @@ $check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -default
 
 $path = Get-AnsibleParam -obj $params -name "path" -type "path" -failifempty $true -aliases "dest","name"
 $state = Get-AnsibleParam -obj $params -name "state" -type "str" -validateset "absent","directory","file","touch"
+$attributes = Get-AnsibleParam -obj $params -name "options" -type "str" -aliases "attributes"
```
Comment (dagwieers) [inline lib/ansible/modules/windows/win_file.ps1]: I really prefer seeing the same parameters as the **file** module, and in this case make **attributes** the parameter, not **options**. (Not sure why this is called options here)

https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/module_docs_fragments/files.py#L75

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
So one of my observations here is that we would like to have diff support in this module at some point, just like the **file** module, and it would be nice if the implementation at least makes that still possible (rather than harm such effort in the future).

I know I cannot ask you to add diff-support when you're extending a module without it, but sure would be nice ;-)

**Comment by samdoran** [comment]
cc @jborean93 @nitzmahone

**Comment by jhawkesworth** [comment]
Hi,

Thank you for this, I think its a good idea to have the ability to modify file attributes,
however I have a few suggestions.

I suggest using a list for $attributes rather than a str

I don't think we currently have a '-validatelist' function in Ansible.ModuleUtils.Legacy.psm1
but it would be useful for something like this.  

I think from the code L83-101 you are going for a something like attrib +r / attrib -s style arguments?
While this is I guess familiar, ansible is strong on human readability
so I'd prefer if the attribute names were spelled out in full

"archive", hidden", "readonly", "system" and you could list whatever you need applying in your playbook like this

attributes:
  - hidden
  - system

or 

attributes: [archive, readonly]

'normal' would then just be 
attributes: []
or simply don't specify the attributes: parameter in the playbook.

That way you are stating what state you need the file to be in and ansible just ensures it is as the playbook expects (I find
the + or - syntax reminiscent of a 'perform this operation' syntax rather than 'ensure this state' which to me fits more naturally with ansible).

Also if the module expects these names rather than a/h/r/s single character
abbreviations, another advantage would be you would need less complex
logic to translate between what comes from the playbook and the
System.IO.FileAttributes]:: constants in the first part of Set-FileAttributes

A minor nit: consistent naming style for variables would be nice
eg current_attributes instead of CurrentAttributes
personally I don't mind snake_case or CamelCase for variables but like consistency within a chunk of code.

Please update documentation (win_file.py) with examples.

Also please consider adding integration tests to demonstrate the functionality.
See test/integration/targets/win_file/ for the existing tests (which you can run for yourself against your own windows machines using `ansible-test windows-integration win_file` )

L103->L122 could probably be a function with the a/r/s/h  and [System.IO.FileAttributes]::<type> passed into it, which would ease reading and reduce the number of conditionals (if and elses) in the function.

L136 inconsistent indentation (readability)

Please put win_stat changes into a different PR (and please also document win_stat changes).
I know this seems picky but its easier to review smaller changes (and easier to back out  / cherry pick) if necessary.
If you mention the other PR github does a nice job of linking to it so its easy to communicate that the two changes should 'travel together'.

Also consider moving the function into the shared code area ( lib/ansible/module_utils/powershell/ )
I'd suggest in a new file called something like Ansible.ModuleUtils.FileAttributes.psm1
By putting the logic in shared code  win_copy could (in a different PR) be modified to use the same logic as win_file
( much in the same way that copy and file modules can both set file modes in the same way).
To drag in shared code just use `#Requires <name of .psm1 file>` - the facility is new but you can see it being used in win_copy.ps1 right now.

Sorry, quite a lot of feedback but this is good stuff and it would be great to add this functionality.
Happy to chat on #ansible-windows irc channel if you want to discuss more.

**Comment by dagwieers** [comment]
Beware that the deadline for getting new features/modules accepted in [Ansible v2.6 is nearing](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_6.rst#release-schedule), it is set to either 2018-05-17 or 2018-05-25. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to [the Windows Working Group meeting agenda](https://github.com/ansible/community/issues/294) to get it resolved.

**Comment by dagwieers** [comment]
Beware that the deadline for getting new features/modules accepted in [Ansible v2.7 is nearing](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_7.rst#release-schedule), it is set to either 2018-08-23 or 2018-08-30. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to the [Windows Working Group meeting agenda](https://github.com/ansible/community/issues/294) to get it resolved.

**Comment by jhawkesworth** [comment]
Just looking through some old PRs.  I hope my very long comment didn't scare you off @chrisjohnson1 
win_file has moved on, although it still doesn't have the features you were proposing.
Do you have time to work on this at all, or should we close this PR?
Many thanks,
Jon

**Comment by jborean93** [comment]
I do apologise for the delay in getting to this one. We are trying to improve our response times in getting to these PRs but obviously we have a lot of work to try and improve on. Due to the conflict with existing files and the age of the PR I am going to close this PR. This is not to say the work here isn't needed in Ansible so if you are still willing to contribute the changes please open a new PR with the conflicts resolved and we can work on it together.

**Pull request closed without merging**